### PR TITLE
[1.11] Allow any Item to disable a Sheild

### PR DESCRIPTION
**Pull request by Madsthunder**
This allows items other than those that extend ItemAxe to bypass a Sheild.

### CODE REVIEW

**Comment by williewillus** [comment]
I would call it "canDisableShield"

**Comment by mezz** [comment]
This looks good. 
I agree with williewillus about the name, please change it to that.

**Comment by 61352151511** [comment]
Well not completely understanding the functionality I would say it should be canBypassShield if it's just bypassing the shield and canDisableShield if it actually disables it. Disabling would be a permanent effect like you hit the shield with it once and then everything else is effective against it.

**Comment by williewillus** [comment]
No, i suggested that because the feature that your if statement is guarding is literally called "shield disabling" by the vanilla community.

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 1]
```diff
@@ -669,12 +669,17 @@
 +        return builder.build();
 +    }
 +
++    public boolean canBypassSheild(ItemStack stack, ItemStack sheild, Entity entity, Entity attacker)
```
Comment (williewillus) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: shield is spelled "shield" not "sheild" (method name and second param)

**Comment by Madsthunder** [comment]
Added a few commits that adds a Javadoc, corrects the spelling of 'sheild', and renames Item#canBypassSheild to #canDisableShield.

**Comment by williewillus** [comment]
looks good to me

File: patches/minecraft/net/minecraft/entity/monster/EntityMob.java.patch [hunk 2]
```diff
@@ -0,0 +1,11 @@
+--- ../src-base/minecraft/net/minecraft/entity/monster/EntityMob.java
++++ ../src-work/minecraft/net/minecraft/entity/monster/EntityMob.java
+@@ -120,7 +120,7 @@
+                 ItemStack itemstack = this.func_184614_ca();
+                 ItemStack itemstack1 = entityplayer.func_184587_cr() ? entityplayer.func_184607_cu() : ItemStack.field_190927_a;
+ 
+-                if (!itemstack.func_190926_b() && !itemstack1.func_190926_b() && itemstack.func_77973_b() instanceof ItemAxe && itemstack1.func_77973_b() == Items.field_185159_cQ)
++                if (!itemstack.func_190926_b() && !itemstack1.func_190926_b() && itemstack.func_77973_b().canDisableShield(itemstack1, itemstack, p_70652_1_, this) && itemstack1.func_77973_b() == Items.field_185159_cQ)
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/monster/EntityMob.java.patch]: What is `Items.field_185159_cQ`?

Comment (williewillus) [reply patches/minecraft/net/minecraft/entity/monster/EntityMob.java.patch]: that's probably testing for the shield itself. that part isn't being changed in this PR

Comment (mezz) [reply patches/minecraft/net/minecraft/entity/monster/EntityMob.java.patch]: Yes it's `Items.SHIELD`

Comment (liach) [reply patches/minecraft/net/minecraft/entity/monster/EntityMob.java.patch]: Refer to #3557, clean up needed if both are merged. Okay for now.

**Comment by Madsthunder** [comment]
Any updates on this or #3557?

File: patches/minecraft/net/minecraft/item/Item.java.patch [hunk 3]
```diff
@@ -669,12 +669,26 @@
...
++     * 
++     * @param stack The ItemStack
++     * @param shield The shield in question
++     * @param entity The Entity holding the shield
++     * @param attacker The Entity holding the item
++     * @return True if this Item can disable the shield in question
++     */
++    public boolean canDisableShield(ItemStack stack, ItemStack shield, Entity entity, Entity attacker)
++    {
++    	return this instanceof ItemAxe;
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/item/Item.java.patch:683]: tabs eww

File: patches/minecraft/net/minecraft/entity/monster/EntityMob.java.patch (hunk 2 above)
Comment (LexManos) [reply patches/minecraft/net/minecraft/entity/monster/EntityMob.java.patch]: This is probably one of the rare cases where the two PRs should be merged as a general 'expand shield functionality' pr.

**Comment by Madsthunder** [comment]
~~@LexManos Should I remove the `itemstack.getItem() instanceof ItemAxe` altogether in favor of `itemstack.getItem().canDisableShield()`? (That would also slightly reduce the patch size)~~

Scratch that, I never realized a already did that...

**Comment by williewillus** [comment]
patch size is counted by lines added/removed, not how many characters you change in a line.

**Comment by liach** [comment]
@Madsthunder Your suggestion can cause a stack overflow.

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch [hunk 4]
```diff
@@ -148,19 +148,20 @@
 +            this.spawnChunkMap.put(spawndim, new BlockPos(spawndata.func_74762_e("SpawnX"), spawndata.func_74762_e("SpawnY"), spawndata.func_74762_e("SpawnZ")));
 +            this.spawnForcedMap.put(spawndim, spawndata.func_74767_n("SpawnForced"));
 +        }
++        this.spawnDimension = p_70037_1_.func_74767_n("HasSpawnDimensionSet") ? p_70037_1_.func_74762_e("SpawnDimension") : null;
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: Why did you add this?

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch [hunk 5]
```diff
@@ -181,26 +182,39 @@
 +        }
 +        p_70014_1_.func_74782_a("Spawns", spawnlist);
 +
++        p_70014_1_.func_74757_a("HasSpawnDimensionSet", this.hasSpawnDimension());
++        if (this.hasSpawnDimension())
++            p_70014_1_.func_74768_a("SpawnDimension", this.getSpawnDimension());
++
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: Bad addition

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch [hunk 6]
```diff
@@ -589,6 +603,12 @@
 +        return capability == net.minecraftforge.items.CapabilityItemHandler.ITEM_HANDLER_CAPABILITY || super.hasCapability(capability, facing);
 +    }
 +
++    @Nullable
++    private Integer spawnDimension;
++    public boolean hasSpawnDimension() { return spawnDimension != null; }
++    public int getSpawnDimension() { return spawnDimension != null ? spawnDimension : 0; }
++    public void setSpawnDimension(@Nullable Integer dimension) { this.spawnDimension = dimension; }
++
```
Comment (liach) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: Bad addition

**Review by liach — CHANGES_REQUESTED** [review]

**Comment by liach** [comment]
You can try to use `git mergetool`

**Review by liach — APPROVED** [review]

**Comment by Madsthunder** [comment]
@jaredlll08 Would you mind if I merged the content of your pr with this to avoid more merge conflicts?

**Comment by jaredlll08** [comment]
Go ahead :)

**Comment by Madsthunder** [comment]
Wow, I screwed up badly on my attempt to resolve merge conflicts...

**Comment by mezz** [comment]
Closing in favor of #3741

**Pull request closed without merging**